mne/preprocessing/asr_utils.py

```diff
@@ -0,0 +1,509 @@
+    for b in range(len(shape_range)):
+        gam = gammaincinv(
+            1 / shape_range[b], np.sign(quants - 1 / 2) * (2 * quants - 1))
+        zbounds.append(np.sign(quants - 1 / 2) * gam ** (1 / shape_range[b]))
+        rescale.append(shape_range[b] / (2 * gamma(1 / shape_range[b])))
+
+    # determine the quantile-dependent limits for the grid search
+    # we can generally skip the tail below the lower quantile
+    lower_min = np.min(quants)
+    # maximum width is the fit interval if all data is clean
+    max_width = np.diff(quants)
+    # minimum width of the fit interval, as fraction of data
+    min_width = min_clean_fraction * max_width
+
+    # Build quantile interval matrix
+    cols = np.arange(lower_min,
+                     lower_min + max_dropout_fraction + step_sizes[0] * 1e-9,
+                     step_sizes[0])
+    cols = np.round(n * cols).astype(int)
+    rows = np.arange(0, int(np.round(n * max_width)))
+    newX = np.zeros((len(rows), len(cols)))
+    for i, c in enumerate(range(len(rows))):
+        newX[i] = X[c + cols]
+
+    # subtract baseline value for each interval
+    X1 = newX[0, :]
+    newX = newX - X1
+
+    opt_val = np.inf
+    opt_lu = np.inf
+    opt_bounds = np.inf
+    opt_beta = np.inf
+    gridsearch = np.round(n * np.arange(max_width, min_width, -step_sizes[1]))
+    for m in gridsearch.astype(int):
+        mcurr = m - 1
+        nbins = int(np.round(3 * np.log2(1 + m / 2)))
+        cols = nbins / newX[mcurr]
+        H = newX[:m] * cols
+
+        hist_all = []
+        for ih in range(len(cols)):
+            histcurr = np.histogram(H[:, ih], bins=np.arange(0, nbins + 1))
+            hist_all.append(histcurr[0])
+        hist_all = np.array(hist_all, dtype=int).T
+        hist_all = np.vstack((hist_all, np.zeros(len(cols), dtype=int)))
+        logq = np.log(hist_all + 0.01)
+
+        # for each shape value...
+        for k, b in enumerate(shape_range):
+            bounds = zbounds[k]
+            x = bounds[0] + np.arange(0.5, nbins + 0.5) / nbins * np.diff(bounds)  # noqa:E501
+            p = np.exp(-np.abs(x) ** b) * rescale[k]
+            p = p / np.sum(p)
+
+            # calc KL divergences
+            kl = np.sum(p * (np.log(p) - logq[:-1, :].T), axis=1) + np.log(m)
+
+            # update optimal parameters
+            min_val = np.min(kl)
+            idx = np.argmin(kl)
+            if min_val < opt_val:
+                opt_val = min_val
+                opt_beta = shape_range[k]
+                opt_bounds = bounds
+                opt_lu = [X1[idx], X1[idx] + newX[m - 1, idx]]
+
+    # recover distribution parameters at optimum
+    alpha = (opt_lu[1] - opt_lu[0]) / np.diff(opt_bounds)
+    mu = opt_lu[0] - opt_bounds[0] * alpha
+    beta = opt_beta
+
+    # calculate the distribution's standard deviation from alpha and beta
+    sig = np.sqrt((alpha ** 2) * gamma(3 / beta) / gamma(1 / beta))
+
+    return mu, sig, alpha, beta
+
+
+def yulewalk(order, F, M):
```

Comment:
there is already a yule walker code in https://github.com/mne-tools/mne-python/blob/main/mne/time_frequency/ar.py

Reply:
Great, Thanks for pointing that out!